api/src/main/java/org/openmrs/module/fgh/mpi/MpiHttpClient.java

```diff
@@ -122,19 +168,110 @@ public void submitPatient(String patientData) throws Exception {
+			
+			if (log.isDebugEnabled()) {
+				log.debug("MPI patient submission response: " + santeResponse);
+			}
+			
+			if (santeResponse == null || santeResponse.isEmpty()) {
+				throw new APIException("An empty response was received when the patient was submitted");
+			}
+		} else {
+			List<Map<String, Object>> mpiIdsResp = submitRequest(SUBPATH_PATIENT, patientData, List.class);
+			
+			if (log.isDebugEnabled()) {
+				log.debug("MPI patient submission response: " + mpiIdsResp);
+			}
+			
+			if (CollectionUtils.isEmpty(mpiIdsResp) || mpiIdsResp.get(0) == null) {
+				throw new APIException("An empty response was received when the patient was submitted");
+			}
 		}
 		
 		log.info("Successfully submitted the patient record to the MPI");
 	}
 	
+	private <T> void retriveAccessToken(Class<T> responseType) throws MalformedURLException, IOException {
+		
+		AdministrationService adminService = Context.getAdministrationService();
+		
+		String clientId = adminService.getGlobalProperty(GP_SANTE_CLIENT_ID);
+		String clientSecret = adminService.getGlobalProperty(GP_SANTE_CLIENT_SECRET);
+		//String loginType = adminService.getGlobalProperty(GP_SANTE_LOGIN_TYPE);
+		//String scope = adminService.getGlobalProperty(GP_SANTE_SCOPE);
+		
+		String data = "grant_type=client_credentials" + "&" + "scope=*" + "&" + "client_secret=" + clientSecret + "&"
+		        + "client_id=" + clientId;
+		
+		// Request a Refresh token in case it expires 
+		if (this.tokenInfo != null) {
+			
+			if (!this.tokenInfo.isValid()) {
+				//Implement a refresh token method
+				data = "grant_type=refresh_token&refresh_token=" + this.tokenInfo.getRefresh_token() + "&" + "client_secret="
+				        + clientSecret + "&" + "client_id=" + clientId;
+				this.doAuthentication(data);
+			}
+		} else {
+			// Normal Login 
+			this.doAuthentication(data);
+		}
+	}
+	
+	private void doAuthentication(String data) throws MalformedURLException, IOException {
+		AdministrationService adminService = Context.getAdministrationService();
+		
+		String uri = "/auth/oauth2_token";
+		this.serverBaseUrl = adminService.getGlobalProperty(GP_MPI_BASE_URL);
+		String url = serverBaseUrl + uri;
+		
+		HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
+		
+		Range<Integer> successRange = Range.between(200, 299);
+		
+		try {
+			connection.setRequestProperty("Content-type", "application/x-www-form-urlencoded");
+			connection.setRequestProperty("Content-Length", Integer.toString(data.getBytes().length));
+			
+			connection.setConnectTimeout(30000);
+			connection.setUseCaches(false);
+			connection.setDoInput(true);
+			connection.setDoOutput(true);
+			connection.setRequestMethod("POST");
+			
+			connection.connect();
+			
+			OutputStream out = connection.getOutputStream();
+			out.write(data.getBytes());
+			out.flush();
+			out.close();
+			
+			if (!successRange.contains(connection.getResponseCode())) {
```

Comment:
Is it true that actually all 200 values are valid?

Reply:
Did you see this comment?

Reply:
Based on the [OpenCR API documentation](https://intrahealth.github.io/opencr-api/#/), for a bundle and single patent POST it is 201 and for the look up it is 200, but these could be based on the latest version, the version we are using that I coded against I recall they were all 200

Reply:
Anyway this code is ONLY executed for Sante and never for OpenCR, so I can ignore this